### ISSUE TITLE
Display frozen, mouse and keyboard input still works

### ISSUE THREAD

**Issue by tuanpham-dev** [gujjwal00/avnc#92]
I sometimes got this problem when pressing Ctrl+C to copy text especially with this tigervnc option `-ImprovedHextile 0`. See the attachment [log.txt](https://github.com/gujjwal00/avnc/files/9446764/log.txt)

**Comment by gujjwal00** [comment]
I tried with two different versions of tigervnc (1.11 & 1.7), but could not reproduce this after few tries. Are you able to reproduce this every time? I assume you were pressing Ctrl+C on the server? 

If only frame is frozen and input is working, then the receiver thread might be stuck on something. Can you please add debug logs in [message processing loop](https://github.com/gujjwal00/avnc/blob/1fe3dedc5281db3d6889dfde15ead331c4bf9e62/app/src/main/java/com/gaurav/avnc/viewmodel/VncViewModel.kt#L228-L229) to detect if receiver thread is running during a freeze. Something like this:

```kotlin
    private fun processMessages() {
        while (viewModelScope.isActive) {
            Log.d(javaClass.simpleName, "Processing message")
            client.processServerMessage()
        }
        Log.d(javaClass.simpleName, "Processing finished")
    }
```

**Comment by tuanpham-dev** [comment]
For more context, I am using linux deploy for the server on the same device. The chance I got the problem was about 30%.

After enable developer options I rarely see the problem (not sure why). I am closing this issue and will dig into the problem when I have time. Will let you know if I find anything.

**Comment by tuanpham-dev** [comment]
This issue is happening again to me after 1-2 hours using the app for connecting local chroot server. I got this error in logs `OpenGLRenderer: setSurface called with nullptr`. I think it is related to [this](https://bugzilla.mozilla.org/show_bug.cgi?id=1762424#c3).

**Comment by gujjwal00** [comment]
Is there anything special about when this issue occurs, aside from `Ctrl+C`? E.g. device is rotated, server resolution is changed or memory is low. Is it possible to try a different server to rule out any server-side issue?

I don't have a rooted device to reproduce this, so if you want I can prepare an APK with extra logging to detect deadlocks.

**Comment by tuanpham-dev** [comment]
It occurs after long time of running when memory usage is high (mainly chromium) and my tablet is a little hot. It even occurs when I select text before copying.

Yes, it would be nice if you can provide the apk so I can test myself and give you the logs.

**Comment by gujjwal00** [comment]
I have added some logging to monitor different threads: https://github.com/gujjwal00/avnc/commit/afa86d7185c365d49f4b0ea5e2f16579a9013155
CI built APK: https://github.com/gujjwal00/avnc/suites/8611507064/artifacts/386289451

During VNC session, ThreadMonitor will log some stats every 2 seconds. If any thread is blocked, or server is not sending framebuffer updates, we should be able to identify it.

**Comment by tuanpham-dev** [comment]
Hi, the please check the log attached
[aa28215b.txt](https://github.com/gujjwal00/avnc/files/9731960/aa28215b.txt)

**Comment by gujjwal00** [comment]
Most recent ThreadMonitor stats from the logs are as follows:

```python
10-07 15:01:23.155 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.148s]  [Renderer: 5.823s]  [Sender: 5.823s]  [Receiver-FBU: 7.164s] 
10-07 15:01:25.158 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.149s]  [Renderer: 7.827s]  [Sender: 7.827s]  [Receiver-FBU: 9.168s] 
10-07 15:01:27.160 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.148s]  [Renderer: 9.828s]  [Sender: 9.828s]  [Receiver-FBU: 11.169s] 
10-07 15:01:29.164 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.148s]  [Renderer: 11.833s]  [Sender: 11.833s]  [Receiver-FBU: 13.174s] 
10-07 15:01:31.166 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.139s]  [Renderer: 0.006s]  [Sender: 0.007s]  [Receiver-FBU: 15.175s] 
10-07 15:01:33.168 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.597s]  [Renderer: 0.592s]  [Sender: 0.879s]  [Receiver-FBU: 0.597s] 
10-07 15:01:35.169 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.014s]  [Renderer: 0.002s]  [Sender: 0.007s]  [Receiver-FBU: 0.014s] 
10-07 15:01:37.170 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.001s]  [Renderer: 0.001s]  [Sender: 0.575s]  [Receiver-FBU: 0.001s] 
10-07 15:01:39.172 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.037s]  [Renderer: 0.037s]  [Sender: 0.088s]  [Receiver-FBU: 0.037s] 
10-07 15:01:41.173 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.034s]  [Renderer: 0.032s]  [Sender: 0.315s]  [Receiver-FBU: 0.034s] 
10-07 15:01:43.173 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.0s]  [Renderer: 0.004s]  [Sender: 0.043s]  [Receiver-FBU: 0.0s] 
10-07 15:01:45.174 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.0s]  [Renderer: 0.004s]  [Sender: 0.0s]  [Receiver-FBU: 0.0s] 
10-07 15:01:47.174 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.004s]  [Renderer: 0.008s]  [Sender: 0.006s]  [Receiver-FBU: 0.004s] 
10-07 15:01:49.174 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.0s]  [Renderer: 0.007s]  [Sender: 0.004s]  [Receiver-FBU: 0.0s] 
10-07 15:01:51.175 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.108s]  [Renderer: 0.001s]  [Sender: 0.001s]  [Receiver-FBU: 0.108s] 
10-07 15:01:53.175 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.321s]  [Renderer: 0.008s]  [Sender: 0.008s]  [Receiver-FBU: 0.321s] 
10-07 15:01:55.175 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.92s]  [Renderer: 0.006s]  [Sender: 0.004s]  [Receiver-FBU: 0.924s] 
10-07 15:01:57.177 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.92s]  [Renderer: 0.351s]  [Sender: 0.353s]  [Receiver-FBU: 2.926s] 
10-07 15:01:59.178 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.919s]  [Renderer: 0.003s]  [Sender: 0.004s]  [Receiver-FBU: 4.927s] 
10-07 15:02:01.180 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.918s]  [Renderer: 0.088s]  [Sender: 0.087s]  [Receiver-FBU: 6.929s] 
10-07 15:02:03.180 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.915s]  [Renderer: 0.059s]  [Sender: 0.059s]  [Receiver-FBU: 8.929s] 
10-07 15:02:05.182 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.914s]  [Renderer: 0.004s]  [Sender: 0.01s]  [Receiver-FBU: 10.931s] 
10-07 15:02:07.183 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.912s]  [Renderer: 0.011s]  [Sender: 0.011s]  [Receiver-FBU: 12.931s] 
10-07 15:02:09.186 30864 30864 D ThreadMonitor: Ping report:  [Receiver: 0.914s]  [Renderer: 2.002s]  [Sender: 0.263s]  [Receiver-FBU: 14.935s] 
10-07 15:02:09.676 30864 30864 D ThreadMonitor: Stopped
10-07 15:05:55.501 30864 30864 D ThreadMonitor: Started
```

Looking at these logs it appears that around `15:01:55` AVNC stopped receiving framebuffer updates from server. You can see the time since last FBU keep increasing, even though all threads are still running. 

I am assuming this is when display appeared frozen, and after about 14 seconds at  `15:02:09` you stopped the connection and started a new session.

**Comment by tuanpham-dev** [comment]
I think the problem is from libvncclient, I have the same problem with multivnc. bvnc and vnc viewer still work fine.

**Comment by gujjwal00** [comment]
Yes, it could be a bug in libvncclient. But given the nature of this bug, it will be pretty hard to debug this.

Just to rule out a server issue, have you tried running a different server?

**Comment by gujjwal00** [comment]
@tuanpham-dev have you tried the workaround mentioned in #105, i.e. disabling battery optimizations

**Comment by tuanpham-dev** [comment]
@gujjwal00 - Yes I did. Haven't had time to use this app long enough now but seeing the client is still responsive after sleep for hours.

**Comment by tuanpham-dev** [comment]
Now it turns back to normal after few seconds delay when the display is frozen. I think this issue can be closed. My only hope is there is a way to configure the timeout to shorten the delay. But it is ok if 5 seconds is required or the settings doesn't get bloated.

**Comment by gujjwal00** [comment]
I am not sure which timeout you are running into. I haven't changed anything related to this issue recently.

**Comment by tuanpham-dev** [comment]
This commit https://github.com/gujjwal00/avnc/commit/e109989fc4b9fa0013db08f9c86e8e842ce9fa25 perhaps?

**Comment by gujjwal00** [comment]
Auto-reconnect will only kick in when you are on `Disconnected` screen. It should not affect the the frozen state, because as far as AVNC is concerned, connection is still active.

**Comment by gujjwal00** [comment]
I think the root cause of this bug was fixed in issue #156 . Fix was implemented in LibVNCClient via https://github.com/gujjwal00/libvncserver/commit/b495f1cf6676056b99d3eebc288ce3232f971de0 .

Basically, LibVNCClient got into a situation where it had read the update message into a buffer but the function which checks if update is available didn't look there.

**Comment by tuanpham-dev** [comment]
Nice to see it fixed. Thanks!
